### ISSUE TITLE
chore | backend | POC: recall-safe ranked-list truncation — viability & go/no-go

### ISSUE THREAD

**Issue by AmosChenZixuan** [AmosChenZixuan/BibiLab#530]
Depends on: #529 (soft — re-measure separation after query hygiene lands; sharper queries may be what makes a clean cut possible)

## Goal
Decide, **with evidence**, whether to replace `find_passages`' flat top-8 truncation with a recall-safe adaptive cut. This is an INVESTIGATION, not a feature: the deliverable is a written go/no-go plus the data behind it. Shipped truncation code is a SEPARATE follow-up feature, filed only if this POC greenlights it.

## Background / evidence already gathered
- `retrieve()` (`backend/src/bibilab/pipeline/embed.py:522`) returns `chunks[:top_k]` with `top_k = FIND_PASSAGES_TOP_K = 8`, **gateless** — no relevance floor ("rerank is ordering, not authority"). Candidate pool is `effective_top_k = min(max(sources_total*3, top_k, 10), 60)` (`embed.py:537`); the final cut is `chunks[:top_k]` (`embed.py:556`).
- Measured waste (join of rendered fences in `~/.bibilab/debug/{msg_id}.json` against persisted `messages.metadata.rag.calls[].context[]` cited sections in `~/.bibilab/bibilab.db`, n=11 turns): **59% of surfaced sections are never cited and never read** (upper bound; "used" = cited∪read is itself a lower bound on usefulness).
- Regime is **bimodal**: concentrated turns surface 6-8, use 1-2 (huge wasteful tail); enumeration turns use 5/5 (a cut there loses recall). One policy must serve both.
- Cited-section rerank scores span **−0.53 → 8.09** (mean 2.20). Used sections are NOT reliably high-scoring, so a naive cut risks dropping used sections. **Whether a SAFE cut exists is the open question this POC answers.**

## Why "instrument" and "cut" are merged here
The instrumentation has no standalone deliverable — it exists only to produce the separation data that decides the cut, and the cut is not a feature until that data says yes. So the unit of work is: instrument -> measure -> evaluate -> decide.

## What to do
1. **Instrument (throwaway scaffolding, NOT a feature).** The unused tail's rerank scores are currently recorded *nowhere* — `context[]` keeps only cited sections (built at `chat_tools.py:586`; per-entry score set at `chat_tools.py:523`). Capture ALL surfaced sections' `rerank_score` + a `cited` flag. Two seams: surfaced+scores at `find_passages` execution (`execute_find_passages`, `chat_tools.py:352`); cited-set joined post-stream at the terminal `rag` reconstruction in `run_chat_turn` (`backend/src/bibilab/routers/chat.py`, ~lines 700-720). **DECISION POINT (pick at dispatch):** extend the persisted `rag` metadata to keep the full surfaced set, vs. write a separate retrieval-trace JSONL. This logging is REMOVED when the POC concludes (or promoted by the follow-up feature) — do not leave permanent dead instrumentation.
2. **Collect data. DECISION POINT (recommend ORGANIC):** open the log, use the app normally, accumulate traces, analyze when enough — no heavy batch run. Alternative: replay recorded queries through live retrieval (heavy — loads the ONNX reranker + embedder, so **concurrency 1 / foreground / never backgrounded**, per project constraint). MUST include >=1 NON-contaminated list: the 诡秘之主 test list (`3cca7e71`) is parametric-contaminated, so its grounding/citation signal is unreliable for this measurement.
3. **Analyze (throwaway script — no production hooks, no eval-framework dressing).** Produce a used-vs-unused rerank-score ROC / PR curve. Evaluate candidate cut policies: score-distribution mixture crossover (Arampatzis–Kamps–Robertson, "Where to stop reading a ranked list"), relative-to-top band, simple quantile. Report **per regime** (concentrated vs enumeration): recall retained vs % tail removed.
4. **Decide against the gate and write the go/no-go.**

## Go/no-go gate (DECISION POINT — defaults below, confirm at dispatch)
- **GREENLIT** iff a cut policy exists with **recall >= 95% overall AND 100% on enumeration-regime turns**, while removing **>= 50% of the unused tail**.
- **NO-GO** if every policy hitting the waste target drops used sections past the recall floor (scores overlap) -> recommendation becomes "ship query hygiene (#529) only; accept the token cost; keep gateless." **No-go is a valid, expected outcome, not a failure.**

## Don't-undo constraints (carry into any follow-up impl)
- Gateless top-k is a frozen "don't-undo" decision (RAG v2). This POC is the deliberate, evidence-backed reopening of it — do not quietly reintroduce a gate without the data.
- Any cut MUST be RELATIVE, never a fixed score threshold: bge-reranker-base scores are uncalibrated logits (can be negative); a fixed threshold is exactly what was deleted (`backend/src/bibilab/pipeline/rerank.py:24-25`). Reranker calibration (Platt / isotonic) is the canonical alternative that would make a fixed threshold interpretable — note it as considered-but-needs-labeled-dev-set.
- Fail-open / recall-biased: when the distribution is flat/unimodal (enumeration), keep everything up to the ceiling.
- CRAG-style retrieval-evaluator critics are OUT (razor stance: prefer LLM judgment / use the already-computed signal; add no critic model).

## Out of scope
- Shipping the truncation into `retrieve()` — that is the follow-up `feat` issue, filed only if greenlit.
- The query-hygiene prompt change (#529).
- Permanent observability / eval-framework integration for the score log.

## Acceptance (deliverable = a decision, not shipped code)
- [ ] Score-capture scaffolding records every surfaced section's `rerank_score` + `cited` flag for live turns.
- [ ] An analysis over >=1 non-contaminated list produces the used-vs-unused score curve + a per-regime recall/waste table.
- [ ] A written go/no-go is posted to this issue, evaluated against the gate, with the numbers.
- [ ] If GREENLIT: a follow-up `feat | backend | recall-safe ranked-list truncation` issue is filed with the chosen policy + thresholds. If NO-GO: the scaffolding is removed and this issue is closed with the evidence.

**Comment by AmosChenZixuan** [comment]
## Go / No-Go: **NO-GO**

A recall-safe relative cut that removes ≥50% of the waste **does not exist** on bge-reranker-base scores for this corpus. Keep the gateless top-k. Ship #529 (merged in #531), accept the token cost.

### Method
Replayed **14 user messages** mined from all 32 prompt-trace dumps in `~/.bibilab/debug/`, spanning 3 corpora — 黑塔 (fantasy audio-novel), 美食 (cooking), 诡秘之主 (`3cca7e71`, **parametric-contaminated → excluded from the gate**, reported for reference) — and both regimes (concentrated / enumeration / recommendation), through the **post-#529** chat path. For every `find_passages`-surfaced section captured its bge `rerank_score` + a `used` flag (`cited ∪ read`). Self-contained throwaway harness, foreground / concurrency-1. **83 surfaced sections (73 on the clean corpora).**

Waste reproduced: **38/83 (46%) surfaced sections never used** (clean-only 34/73 = 47%) — same order as the prior n=11 / 59% estimate.

### Used-vs-unused score separation (clean corpora)
| set | n | min | median | max | mean |
|---|---|---|---|---|---|
| USED | 39 | −1.06 | 2.61 | 8.56 | 2.53 |
| UNUSED | 34 | −6.34 | −1.03 | **5.12** | −0.51 |

The distributions overlap heavily: unused scores reach **5.12**, used scores fall to **−1.06**. No global threshold separates them.

### Relative-band cut `keep iff score ≥ turn_top − δ` (the only legitimate form — bge logits are uncalibrated)
Gate: recall ≥95% overall **AND** 100% on enumeration **AND** waste-removed ≥50%.

| δ | CLEAN recall / waste | enumeration recall / waste |
|---|---|---|
| 2.0 | 67% / **56%** | 65% / 75% |
| 3.0 | 79% / 44% | 83% / 50% |
| 4.0 | **95%** / 38% | **100%** / 38% |
| 5.0 | 100% / 38% | 100% / 38% |

**No δ satisfies the gate.** Recall ≥95% forces waste-removed ≤38%; waste ≥50% forces recall ≤67%. The two requirements never hold together.

### Why — structural score inversion (the decisive finding)
**6 of 14 turns have `max(unused score) > min(used score)`** — the reranker ranks an *unused* section above a *used* one:

| turn | used-min | unused-max |
|---|---|---|
| 归树是谁 (heita, concentrated) | 4.49 | **5.12** |
| 念刃是什么 (heita, concentrated) | 1.50 | **2.61** |
| 我想吃鸡肉和面条…推荐 (food, recommendation) | 1.52 | **2.42** |
| 有哪些用鸡蛋做的菜 (food, enumeration) | −0.73 | **−0.04** |
| 第五纪元的编年史 (mister*) | −1.02 | **3.24** |

When the ranking itself inverts usefulness, **any monotonic cut** (relative band *or* a calibrated absolute threshold) that drops the unused section also drops a used one. Score is an *ordering*, not an authority signal — exactly the gateless-top-k premise.

### Alternatives considered & rejected
- **Reranker calibration (Platt / isotonic):** monotonic, so it cannot fix inversions. Would make a fixed threshold *interpretable* but not *separating*. Rejected.
- **Query hygiene (#529):** the inversions occur on already-clean entity queries (`归树是谁`, `念刃是什么`), so sharper queries don't dissolve them. (#529 still ships on its own merits.)
- **CRAG-style relevance critic:** out per the razor stance — and it only relocates the cost (an extra judgment pass) rather than reading it off the existing score.

### Caveats (honest bounds)
- **n=14 turns / 83 sections** — directional, not definitive. But the failure mode is *structural* (rank inversions), not a marginal aggregate gap, so more data refines the recall/waste curve without removing inversions.
- `used = cited ∪ read` is a **lower bound** on usefulness → measured waste is an **upper bound**. Even granting heavy under-citing, the inversions remain unless the model systematically failed to cite its highest-scored relevant sections (implausible).
- *mister excluded from the gate (contamination): `有哪些古神` cited only 1/3 surfaced — consistent with answering from parametric canon rather than the transcripts.*

### Outcome
Per the acceptance criteria, NO-GO ⇒ no follow-up `feat` filed; gateless top-k stays; throwaway scaffolding removed. Closing this issue with the evidence above.

**Comment by AmosChenZixuan** [comment]
NO-GO — evidence in the comment above. Gateless top-k retained; #529 (query hygiene) shipped separately in #531. Scaffolding removed.

**Comment by AmosChenZixuan** [comment]
## Correction — the "score inversion" framing was too strong; refined diagnosis

Inspecting the actual section text behind the inversion cases (retrieval-only, query passed directly, relevance judged by reading — not by what the LLM cited) **partly walks back** the "rerank score inversion ⇒ reranker is the problem" reading. The NO-GO stands, but the mechanism is different and the reranker is largely **not** at fault.

### Case A — 炝锅面 (precise query): reranker separates cleanly, no inversion
Re-running with the bare entity query `炝锅面怎么做`:
```
[1]  8.52  ← the actual 炝锅面 video
[2] -1.49  ← 懒人一周晚餐 (meal-prep, not 炝锅面)
[3] -2.23  ← 嫩牛五方/煎饼
```
A **cliff from 8.52 to −1.49** — textbook separation. The POC had recorded `[2]=4.64` (high, unused) for this dish, but that was under the **LLM-formulated** query that turn, not the clean entity query. So that particular "inversion" was **query-formulation variance, not a reranker defect**. Given a clean query, the reranker's tail is correctly low.

### Case B — 归树是谁 (entity-saturated corpus): tight high band, not mis-ranking
```
[1] 5.12  [2] 5.12  [3] 4.55  ...  [8] 4.40
```
All 8 sections are **genuinely about 归树** (the protagonist — he's in every episode; one section literally contains the line "归树是谁呢？"). The reranker scoring them all ~equally high is **correct**, not an error. The corpus has **no single "归树 = X" definition section**; identity is woven through the narrative. So the LLM cites one or two representative sections and the rest are **relevant-but-redundant**, not noise. The "5.12 > 4.49" gap is jitter inside a band where everything is genuinely relevant — "used" is a near-arbitrary representative subset.

### Refined conclusion (NO-GO unchanged, cleaner reasons)
- The reranker is **not broken** — it separates well on precise queries (Case A). A bigger/better reranker would **not** change the outcome.
- A large share of the measured "46% waste" is **relevant-but-redundant** sections on topic-homogeneous / entity-saturated corpora (Case B) + **query-formulation variance** + the fact that `used = cited∪read` is a strict **lower bound** on usefulness. Cutting by score would drop **signal, not noise**.
- Net: the truncation premise ("there's a noise tail to remove") is **partly false** — on the hard cases the tail is redundant signal, and on the easy cases the reranker's tail is already low (no aggressive cut needed). Both point the same way: **keep gateless top-k.**

The earlier comment's per-turn "inversion" table over-attributed this to reranker quality; the corpus-redundancy + label-lower-bound + query-variance explanation is the accurate one. Flagging so nobody chases a reranker upgrade as the fix.
